variant/effect.go

```diff
@@ -0,0 +1,276 @@
+	Type       EffectType
+}
+
+// EffectType describes the overall effect of a CodingChange.
+// Possible EffectTypes are described below.
+type EffectType byte
+
+const (
+	Silent           EffectType = 6 // mutation changes codon, but still codes for the same amino acid
+	Frameshift       EffectType = 5 // shifts reading frame
+	Nonsense         EffectType = 4 // premature termination
+	InFrameInsertion EffectType = 3 // insertion where insertedBases % 3 == 0
+	InFrameDeletion  EffectType = 2 // deletion where deletedBases % 3 == 0
+	Missense         EffectType = 1 // mutation changes coded amino acid
+)
+
+// Effector is satisfied by types that implement the Effect method, which returns
+// a CodingChange describing how changes prescribed by the receiver type alter the
+// resulting amino acid sequence corresponding to the input []dna.Base. The input
+// sequence should be the CDS of a gene such that input[0:3] is the start codon.
+// The last 3 dna.Base of the input is not required to be a stop codon. Trailing
+// sequence after the stop codon (i.e. the 3' UTR) may be optionally added to the
+// input sequence, however the deletion position must be within the coding sequence.
+// This trailing sequence is used when a variant causes a frameshift and a new stop
+// codon in not found prior to the original stop codon. In this case the translation
+// will continue into the 3' UTR attempting to find a new stop codon.
+//
+// The offset integer input for the Effect method is added to the position in
+// the receiver type and may be positive or negative. Often this offset will
+// be the negative genomic position of the beginning of the start codon.
+//
+// Effector is satisfied by Substitution, Insertion, Deletion, and Delins.
+// The effects of Structural variation are often too complex to be described by
+// CodingChange and therefore should be handled separately.
+type Effector interface {
+	Effect(codingSeq []dna.Base, offsetStart int, offsetEnd int) (CodingChange, error)
+}
+
+// Effect for Substitution determines a coding change resulting from a substitution.
+func (s Substitution) Effect(codingSeq []dna.Base, offsetStart int, offsetEnd int) (CodingChange, error) {
+	var answer CodingChange
+	offsetPos := s.Pos + offsetStart
+	answer.CodingPos = offsetPos
+	answer.ProteinPos = offsetPos / 3
+
+	if offsetPos < 0 {
+		return answer, ErrNegPos
+	}
+
+	if codingSeq[offsetPos] != s.Ref {
+		return answer, ErrRefMatch
+	}
+
+	frame := offsetPos % 3
+	codonStart := offsetPos - frame
+
+	codon := dna.Codon{codingSeq[codonStart], codingSeq[codonStart+1], codingSeq[codonStart+2]}
+	refAa := dna.GeneticCode[codon]
+
+	codon[frame] = s.Alt // mutate codon
+	altAa := dna.GeneticCode[codon]
+
+	if refAa != altAa {
+		answer.RemovedAa = []dna.AminoAcid{refAa}
+		answer.AddedAa = []dna.AminoAcid{altAa}
+	}
+
+	switch {
+	case altAa == refAa:
+		answer.Type = Silent
+
+	case altAa == dna.Stop:
+		answer.Type = Nonsense
+
+	default:
+		answer.Type = Missense
```

Comment:
Should silent be an option?  I am thinking of mutating a DNA sequence at a site(s) and wanting to know the effect of those mutations

Reply:
The first case checks if the altAa is the same as the refAa. If it is, the effect is reported as Silent.